src/compiler/binder.ts

```diff
@@ -889,7 +889,7 @@ namespace ts {
             return isDottedName(expr)
                 || (isPropertyAccessExpression(expr) || isNonNullExpression(expr) || isParenthesizedExpression(expr)) && isNarrowableReference(expr.expression)
                 || isBinaryExpression(expr) && expr.operatorToken.kind === SyntaxKind.CommaToken && isNarrowableReference(expr.right)
-                || isElementAccessExpression(expr) && isStringOrNumericLiteralLike(expr.argumentExpression) && isNarrowableReference(expr.expression)
+                || isElementAccessExpression(expr) && (isStringOrNumericLiteralLike(expr.argumentExpression) || isIdentifier(expr.argumentExpression)) && isNarrowableReference(expr.expression)

```

Comment:
Does this mean that the following will work

```ts

    
Does that mean something like


import {​​​​​​​​​ key }​​​​​​​​​ from "module";
class C {​​​​​​​​​
  [key]: string;
  constructor() {​​​​​​​​​
    this[key] = "hello";
  }​​​​​​​​​
}​​​​​​​​​
```

but not this?

```ts
import * as mod from "module";
class C {
  [mod.key]: string;
  constructor() {
    this[mod.key] = "hello";
  }
}
```

Reply:
[Confirmed using one of the prior packed builds](https://www.staging-typescript.org/play?ts=4.5.0-pr-45974-3#code/KYDwDg9gTgLgBAYwgOwM7zAGwIYEtkDSwAnnALxwDKxAtgEYSYAUAlANwBQHy2NwqYbAmBwaEACZwA3hzhxQkWIhTo4AaxLkqtBs3YcAvlwXR4CHKlRwAwtNlwA9A7gB3aGtT2A2ljyESALoAXHDoUPgA5pz2TvJQUNDeYuIAdBrEwaEw4chRXHJIaNkArggw0Kx2cnIwABa4qD44+EQZWgBEtcCYmBDt9jX1jclpgR1dPX2cckYGQA):

```ts
export const plainKey = Symbol();

namespace mod {
  export const key = Symbol();
}

export class C {
  // works
  [plainKey]: string;

  // error
  [mod.key]: string;

  constructor() {
    this[plainKey] = "hello"
    this[mod.key] = "hello";
  }
}
```

Reply:
@DanielRosenwasser Yes. Not sure, should `PropertyAccessExpression` be allowed as well?

Reply:
I feel like any access expression should be valid (so both PropertyAccess and ElementAccess)

Reply:
You want probably `isEntityNameExpression`, as used by `isLateBindableName` (which, depending on structure, may be appropriate instead).